components/push/ffi/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+) -> u64 {
+    log::debug!("push_connection_new {} {} -> {} {}=>{}",
+        socket_protocol, server_host, bridge_type, sender_id, registration_id);
+    // return this as a reference to the map since that map contains the actual handles that rust uses.
+    // see ffi layer for details.
+    CONNECTIONS.insert_with_result(error, || {
+        let reg_id = ffi_support::rust_string_from_c(registration_id);
+        let host = ffi_support::opt_rust_string_from_c(server_host);
+        let protocol = ffi_support::opt_rust_string_from_c(socket_protocol);
+        let bridge = ffi_support::opt_rust_string_from_c(bridge_type);
+        let sender = ffi_support::rust_string_from_c(sender_id);
+        let key = ffi_support::opt_rust_string_from_c(encryption_key);
+        let db_path = ffi_support::opt_rust_string_from_c(database_path);
+        let config = PushConfiguration{
+            server_host: host,
+            http_protocol: protocol,
+            bridge_type: bridge,
+            registration_id: reg_id,
+            sender_id: sender,
+            database_path: db_path,
+            .. default()
+        };
+        connect(config)
+    })
+}
+
+// Add a subscription
+/// Errors are logged.
+#[no_mangle]
+pub unsafe extern "C" fn push_get_subscription_info(
+    handle: u64,
+    channel_id: *const c_char,
+    error: &mut ExternError,
+) -> *mut c_char{
+    log::debug!("push_get_subscription");
+    CONNECTIONS.call_with_result_mut(error, handle, |conn| {
+        let options = conn.options.clone();
+        let channel = ffi_support::rust_str_from_c(channel_id);
+        let key = options.vapid_key;
+        let reg_token = options.registration_id;
+        let subscription_key = crypto::generate_key().unwrap();
+        let auth = base64::encode_config(
+            &crypto::get_bytes(SER_AUTH_LENGTH), base64::URL_SAFE_NO_PAD);
+        // Don't auto add the subscription to the db.
+        // (endpoint updates also call subscribe and should be lighter weight)
+        let info = conn.subscribe(channel).unwrap();
+        // store the channelid => auth + subscription_key
+        let mut record = PushRecord::new(
+            &info.uaid,
+            &channel,
+            &info.endpoint,
+            "",
+            subscription_key.clone());
+        record.app_server_key = options.vapid_key.clone();
+        record.native_id = Some(reg_token);
+        self.database.put_record(&record)?;
+        let subscription_info = json!({
+            "endpoint": info.endpoint
+            "keys": {
+                "auth": auth,
+                "p256dh": base64::encode_config(subscription_key.public,
+                                                base64::URL_SAFE_NO_PAD)
+            }
+        });
+        return subscription_info.to_string()
+    })
+}
+
+// Unsubscribe a channel
+#[no_mangle]
+pub unsafe extern "C" fn push_unsubscribe(
+    handle: u64,
+    channel_id: *const c_char,
+    error: &mut ExternError,
+) -> bool{
```

Comment:
Don't return bool over FFI, JNA has bugs around it. I'm actually surprised this works at all, since IntoFfi prevents it.

 (actually, looking more closely, I suspect this doesn't compile at all, since your `call_with_result_mut` callback returns `()` due to the semicolon).

This is documented in the IntoFfi docs, but I'm working on another doc that spells out what you need to know more specifically.